src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -841,6 +832,42 @@ internal bool TryGetCollectionIterationType(ExpressionSyntax syntax, TypeSymbol
             {
                 Error(diagnostics, ErrorCode.ERR_CollectionExpressionTargetTypeNotConstructible, node.Syntax, targetType);
             }
+
+            return;
+
+            NamedTypeSymbol? findSingleIEnumerableTImplementation()
+            {
+                ImmutableArray<NamedTypeSymbol> allInterfaces;
+                switch (targetType.TypeKind)
```

Comment:
>switch (targetType.TypeKind) [](http://example.com/codeflow?start=16&length=28)

The calculation of all interfaces is in three different methods. We should extract a helper method.

And in fact, there is a bug in the original implementation that we're not considering interfaces from the *effective base class* for type parameters. See `ListConstruction_DerivedType_02` from [https://github.com/dotnet/roslyn/pull/70197](https://github.com/dotnet/roslyn/pull/70197) for an example (also below) that should compile successfully:
```csharp
using System.Collections.Generic;
class Program
{
    static U F<T, U>(T x, T y, T z) where U : List<T>, new() => [x, y, z];
}
```

Consider adding and using the `GetAllInterfacesOrEffectiveInterfaces()` extension method from [https://github.com/dotnet/roslyn/pull/70197](https://github.com/dotnet/roslyn/pull/70197).

Reply:
I'd debated doing that, but this implementation throws in the unreachable cases. I'll go ahead with your suggestion. Thanks